### PR TITLE
feat(Table): Single Selectable Rows

### LINKED ISSUE #509: Table: selectable rows without checkbox column
Several people already asked for selectable rows inside Table but without checkbox column when you click on the row, it is being selected. 
Use case: having Table of elements and selecting the row, selects corresponding element in the viewer.
Related #155

### PR DESCRIPTION
**Pull request by adamhock**
- Table now has a prop `selectionMode?: 'multi' | 'single'` which defaults to `multi`. With `single`, we hide the checkbox column so that users can only select a row by clicking on it.

Closes #509

### CODE REVIEW

File: src/core/Table/Table.tsx [hunk 1]
```diff
@@ -121,6 +121,11 @@ export type TableProps<
    * Handler for when a row is clicked. Must be memoized.
    */
   onRowClick?: (event: React.MouseEvent, row: Row<T>) => void;
+  /**
+   * Modify the selection mode of the table
```
[the remark this answers is not in the document]
Comment (veekeys) [reply src/core/Table/Table.tsx:125]: Why you added this prop in the middle of all the props? Cant find the logic.
I'd say expand the description with adding some more info like `Column with checkboxes will not be present with `single` selection mode` and maybe also something about clicking the row will always select that single row.

File: stories/core/Table.stories.tsx [hunk 2]
```diff
@@ -231,12 +232,13 @@ export const Selectable: Story<Partial<TableProps>> = (args) => {
       isSelectable={true}
       onSelect={onSelect}
       onRowClick={onRowClick}
+      selectionMode={'multi'}
```
[the remark this answers is not in the document]
Comment (veekeys) [reply stories/core/Table.stories.tsx:235]: We usually use strings versus objects, so
`selectionMode='multi'`

File: src/core/Table/Table.tsx [hunk 3]
```diff
@@ -121,6 +121,11 @@ export type TableProps<
...
+   * @default 'multi'
+   */
+  selectionMode?: 'multi' | 'single';
```
Comment (veekeys) [inline src/core/Table/Table.tsx:128]: hmm.. I guess we cannot change now, but we could have used just a single prop for enabling selection and setting the mode. just now `isSelectable` name is not really good :/
`selectable?: 'single' | 'multi'`

File: src/core/Table/hooks/useSelectionCell.tsx [hunk 4]
```diff
@@ -7,13 +7,15 @@ import { SelectionColumn, SELECTION_CELL_ID } from '../columns';
...
   isSelectable: boolean,
+  selectionMode: 'multi' | 'single',
   isRowDisabled?: (rowData: T) => boolean,
 ) => (hooks: Hooks<T>) => {
   if (!isSelectable) {
     return;
   }
 
   hooks.allColumns.push((columns: ColumnInstance<T>[]) =>
+    selectionMode === 'single' ||
```
[the remark this answers is not in the document]
Comment (veekeys) [reply src/core/Table/hooks/useSelectionCell.tsx:18]: Personally, I'm fine without comment.

File: src/core/Table/Table.test.tsx [hunk 5]
```diff
@@ -2844,3 +2844,32 @@ it('should show column enabled when whole row is disabled', () => {
...
+  userEvent.click(getByText(mockedData()[1].name));
+  expect(rows[1].classList).toContain('iui-selected');
+  expect(onRowClick).toHaveBeenCalledTimes(1);
+
+  userEvent.click(getByText(mockedData()[2].name));
+  expect(rows[1].classList).not.toContain('iui-selected');
+  expect(rows[2].classList).toContain('iui-selected');
+  expect(onRowClick).toHaveBeenCalledTimes(2);
+
+  userEvent.click(getByText(mockedData()[1].name), { ctrlKey: true });
```
[the remark this answers is not in the document]
Comment (veekeys) [reply src/core/Table/Table.test.tsx:2871]: amm.. No idea, how it should be. Have you tried to check some other examples?

File: src/core/Table/Table.test.tsx [hunk 6]
```diff
@@ -2844,3 +2844,32 @@ it('should show column enabled when whole row is disabled', () => {
...
+
+it('should render selectable rows without select column', () => {
+  const onRowClick = jest.fn();
+  const { container, getByText } = renderComponent({
+    isSelectable: true,
+    selectionMode: 'single',
+    onRowClick,
+  });
+
+  expect(screen.queryByText('Header name')).toBeFalsy();
```
Comment (veekeys) [inline src/core/Table/Table.test.tsx:2856]: Why this is needed?

File: stories/core/Table.stories.tsx [hunk 7]
```diff
@@ -158,7 +159,7 @@ export const Basic: Story<Partial<TableProps>> = (args) => {
   );
 };
 
-export const Selectable: Story<Partial<TableProps>> = (args) => {
+export const SelectableMulti: Story<Partial<TableProps>> = (args) => {
```
[the remark this answers is not in the document]
Comment (veekeys) [reply stories/core/Table.stories.tsx:162]: Local order is weird, at least at my end now :D

image

[the remark this answers is not in the document]
Comment (veekeys) [reply stories/core/Table.stories.tsx:162]: Maybe add them one next to another in this file too?

File: stories/core/Table.stories.tsx [hunk 8]
```diff
@@ -2867,3 +2869,73 @@ ColumnManager.parameters = {
     },
   } as CreeveyStoryParams,
 };
+
+export const SelectableSingle: Story<Partial<TableProps>> = (args) => {
+  const onRowClick = useCallback(
```
[the remark this answers is not in the document]
Comment (veekeys) [reply stories/core/Table.stories.tsx:2874]: Not sure what functionality you have in mind

**Review by veekeys — COMMENTED** [review]
Not a fan of additional prop, but I guess with current props we cannot really just upgrade the `isSelectable`.

File: src/core/Table/Table.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (adamhock) [reply src/core/Table/Table.tsx:125]: My logic was to add this prop at the end of the "Selectable" props (`isSelectable?`, `onSelect?`, `onRowClick?`). Should I move the prop to the end?

File: stories/core/Table.stories.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (adamhock) [reply stories/core/Table.stories.tsx:232]: Changed the local order.

File: stories/core/Table.stories.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (adamhock) [reply stories/core/Table.stories.tsx]: Changed

File: src/core/Table/Table.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (adamhock) [reply src/core/Table/Table.tsx]: Description has been updated

`Modify the selection mode of the table.`
`The Selection Column will not be present with 'single' selection mode.`

File: src/core/Table/Table.test.tsx (hunk 6 above)
Comment (adamhock) [reply src/core/Table/Table.test.tsx]: Found it in the `should handle row clicks` test, but not sure why it's needed. Removed

File: src/core/Table/Table.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (veekeys) [reply src/core/Table/Table.tsx]: amm.. you used capital letters for `Selection Column`. But Im not sure our column is called that way... Maybe just say column with checkboxes or then use the original name of our exported object `SelectionColumn` and wrap it with ` .

[the remark this answers is not in the document]
Comment (adamhock) [reply src/core/Table/Table.tsx]: Changed to "column with checkboxes"

File: src/core/Table/Table.test.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (veekeys) [reply src/core/Table/Table.test.tsx:2870]: Im just thinking.. if user select `single` selection, so it should always be single then, right? ctrl should not even work..
Unless there is case, that users need this multi select behavior, just without checkboxes? (currently, on multi select if you click on rows, it always selects single one)

[the remark this answers is not in the document]
Comment (veekeys) [reply src/core/Table/Table.test.tsx:2870]: @FlyersPh9, maybe you have suggestion from UX perspective here?

[the remark this answers is not in the document]
Comment (adamhock) [reply src/core/Table/Table.test.tsx:2871]: ctrl clicking is disabled in single select mode

**Review by veekeys — APPROVED** [review]

**Review by elephantcatdog — APPROVED** [review]
Looks good to me, just had an unrelated question if anyone knows the answer.

**Pull request merged**